Notifications UI web view2

filipeotero opened the pull request:
### Purpose

This PR implements the Notifications UI that is rendered in a Webview component and its HTML created in React (Which in this PR are included just for sample purposes).

![image](https://user-images.githubusercontent.com/89042471/178281663-cfd3d4c7-0c20-4b92-b702-bfdb3b2bee44.png)

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [X] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@QilongTang @zeusongit @reddyashish 

### FYIs

@jesusalvino @RobertGlobant20

QilongTang commented on src/DynamoCoreWpf/Controls/ShortcutToolbar.xaml:
```diff
@@ -139,6 +140,28 @@
+                                                Foreground="White" 
+                                                FontWeight="Bold" 
+                                                TextAlignment="Center"
+                                                FontSize="9">99
```
Would you add a comment here as TODO to convert this into data binding later?

QilongTang commented on src/DynamoCoreWpf/Controls/ShortcutToolbar.xaml.cs:
```diff
@@ -1,9 +1,12 @@
+using System.Windows.Controls.Primitives;
 using System.Windows.Media;
 using System.Windows.Media.Imaging;
+using Dynamo.Controls;
```
I guess not all of these are needed?

QilongTang commented on src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:
```diff
@@ -95,6 +95,8 @@ public partial class DynamoView : Window, IDisposable
 
         private FileTrustWarning fileTrustWarningPopup = null;
 
+        public ShortcutToolbar ShortcutBar { get { return shortcutBar; } }
```
I think `internal` should be sufficient, if needed for test, we can make it internally visible

filipeotero replied:
I set it to public because it's being used in the notifications extension

reddyashish replied:
Any reason for not making this internal like Aaron suggested?

filipeotero replied:
I'm using it at the notification extension. I didn't think any alternative for it :/

QilongTang commented on src/Notifications/Notifications.csproj:
```diff
@@ -9,8 +9,20 @@
+    <None Remove="View\NotificationUI.xaml" />
+    <None Remove="Web\ArtifaktElement-Regular.woff" />
+    <None Remove="Web\index.bundle.js" />
+    <None Remove="Web\index.html" />
```
This is something we need to work on. I would love to avoid hard copy of the existing component.

RobertGlobant20 commented on src/Notifications/Web/index.html:
```diff
@@ -0,0 +1,20 @@
+<html>
+<head>
+    <meta charset="UTF-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
```
@filipeotero if we are using WebView2 is this line still needed?

filipeotero replied:
not necessary, removed it. Thanks!

RobertGlobant20 commented on src/DynamoCoreWpf/Controls/ShortcutToolbar.xaml:
```diff
@@ -139,6 +140,28 @@
+                                        Background="Transparent">
+                                    <DockPanel>
+                                        <fa:FontAwesome Margin="0,0,20,0"  DockPanel.Dock="Bottom" Icon="BellOutline" Foreground="#38ABDF" FontSize="20px"/>
+                                        <Border DockPanel.Dock="Top" Background="#38ABDF" CornerRadius="2" 
```
I can see that 38ABDF is already defined in DynamoModern.xaml an DynamoColorsAndBrushes.xaml so you can reuse it.

zeusongit commented on src/Notifications/NotificationCenterController.cs:
```diff
@@ -0,0 +1,87 @@
+            var assembly = Assembly.GetExecutingAssembly();
+            var htmlFile = "Dynamo.Notifications.Web.index.html";
+            var fontFile = "Dynamo.Notifications.Web.ArtifaktElement-Regular.woff";
+            var jsFile = "Dynamo.Notifications.Web.index.bundle.js";
```
Can these be exported to resource file? Or are these temporary for testing only?

QilongTang replied:
I think these names should be fixed and they don't need localization. Maybe define as static readonly strings

RobertGlobant20 reviewed (commented):
Some comments

QilongTang commented on src/Notifications/NotificationsViewExtension.cs:
```diff
@@ -98,6 +99,16 @@ public void Loaded(ViewLoadedParams viewStartupParams)
+            var dynamoView = (DynamoView)viewLoadedParams.DynamoWindow;
+            var shortcutBar = dynamoView.ShortcutBar;
+            var notificationsButton = (Button)shortcutBar.FindName("notificationsButton");
+            var notificationCenterController = new NotificationCenterController(dynamoView, notificationsButton);
```
Maybe we can just add the button as a property to the new class? If add button members into the constructor, it might be inconvenient later for new changes

QilongTang commented on src/Notifications/NotificationCenterController.cs:
```diff
@@ -0,0 +1,87 @@
+            notificationUIPopup.PlacementTarget = notificationsButton;
+            notificationUIPopup.Placement = PlacementMode.Bottom;
+            notificationUIPopup.HorizontalOffset = -285;
+            notificationUIPopup.VerticalOffset = 10;
```
These two can be static numbers as well

QilongTang commented on src/Notifications/NotificationCenterController.cs:
```diff
@@ -0,0 +1,87 @@
+            notificationsButton.Click += NotificationsButton_Click;
+
+            notificationUIPopup.webView.EnsureCoreWebView2Async();
+            notificationUIPopup.webView.CoreWebView2InitializationCompleted += WebView_CoreWebView2InitializationCompleted;            
```
We need to unsubscribe these somehow

filipeotero commented:
> @filipeotero Appreciate the implementation!
> 
> 1. Is the bundled js file in this PR your implementation or HIG notification flyout?
> 2. I also believe we are jumping ahead a bit, I was expecting us to implement something like this https://www.meziantou.net/running-npm-tasks-when-building-a-dotnet-project.htm first to have the files for HIG notification flyout and webview2 ready when building Dynamo. FYI: @RobertGlobant20 @jesusalvino

1. It's my implementation
2. Do you mean by this to insert the HIG project on Dynamo and build it with our sln?

QilongTang commented on src/Notifications/NotificationsUIViewModel.cs:
```diff
@@ -0,0 +1,53 @@
+        public double PopupRectangleWidth { get; set; } = 310;
+
+        /// <summary>
+        /// Popup Height, by default it will have a Height of 220
```
The comment here seems wrong or need more explanation about `220`?

reddyashish commented on src/Notifications/NotificationsViewExtension.cs:
```diff
@@ -98,6 +101,14 @@ public void Loaded(ViewLoadedParams viewStartupParams)
+
+        private void LoadNotificationCenter()
+        {
+            var dynamoView = (DynamoView)viewLoadedParams.DynamoWindow;
```
Use: viewLoadedParams.DynamoWindow as DynamoView

reddyashish reviewed (commented):
LGTM with couple of comments.

QilongTang commented on src/Notifications/Web/index.html:
```diff
@@ -0,0 +1,19 @@
+</head>
+<body>
+    <div id="root"></div>
+    <script>#mainJs</script>
```
We may want some comments in this file, maybe can be added into the other repo?

QilongTang commented on src/Notifications/NotificationsUIViewModel.cs:
```diff
@@ -0,0 +1,53 @@
+        /// </summary>
+        public double PopupRectangleHeight { get; set; } = 568;
+                
+        double PointerWidth = 25;
```
Just to confirm, are all of the above properties need to be public? And this one will be internal?

filipeotero replied:
yes, thats right

QilongTang reviewed (commented):
LGTM with some additional comments plus we need the npm install changes (trial npm package by @jesusalvino )

QilongTang commented:
Thank you @filipeotero It looks all comments addressed and we are only missing the npm install step

filipeotero commented:
> Thank you @filipeotero It looks all comments addressed and we are only missing the npm install step

Thank you all for the comments! I entered the npm install step, removed the font style file, and added the base64 version directly in the HTML file. Let me know if it's okay.

zeusongit commented on test/DynamoCoreWpfTests/ViewExtensions/NotificationsExtensionTests.cs:
```diff
@@ -0,0 +1,50 @@
+        }
+
+        [Test]
+        public void ValidateNotificationsUIEmbeededFiles()
```
`ValidateNotificationsUIEmbeededFiles` spell error, update to - > `ValidateNotificationsUIEmbededFiles`

QilongTang commented:
On a second thought `EngOps Build` may fail to build because of the reason @mjkkirschner mentioned that internal Jenkins machine has no access to public npm registry

QilongTang commented:
@filipeotero Do you know why `Dynamo SelfServe` failed?

QilongTang commented:
@filipeotero Can you add the hard copy back or simply revert your recent change for now. Otherwise even the mock would not display because the bundled js does not exist in our repo.
![image](https://user-images.githubusercontent.com/3942418/179663354-e1eec8ff-c09a-4b68-a97e-338eb49775f7.png)
